### PR TITLE
feat(controls): Accordion, migrate knobs to controls - FRONT-1847

### PR DESCRIPTION
**Pull request by planctus**

### CODE REVIEW

**Review by papegaill — APPROVED** [review]

File: src/implementations/twig/storybook/ec/.storybook/main.js [hunk 1]
```diff
@@ -3,12 +3,14 @@ const path = require('path');
 const stories = ['../../../**/!(eu*).story.js'];
 
 const addons = [
+  '@storybook/addon-docs',
```
Comment (papegaill) [inline src/implementations/twig/storybook/ec/.storybook/main.js]: removing this will remove the tabs docs tab

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 2]
```diff
@@ -43,8 +61,16 @@ export default {
...
+Default.argTypes = argTypes;
 Default.storyName = 'default';
-Default.parameters = { notes: { markdown: notes, json: demoData } };
-Default.decorators = [withKnobs, withCode, withNotes];
+Default.parameters = {
+  notes: {
+    markdown: notes,
+    json: demoData,
+  },
+  knobs: {
```
Comment (papegaill) [inline src/implementations/twig/components/accordion/accordion.story.js:67]: do we need to specify this? I guess the addon will be removed globaly

Comment (planctus) [reply src/implementations/twig/components/accordion/accordion.story.js:67]: it will, but then once the migration will be completed in the same ticket we will have to remove the knobs addon we will also remove this from each component that has been migrated to controls. I find it weird to have both panels in place with the knobs one empty, just that...

File: src/implementations/twig/storybook/ec/.storybook/main.js (hunk 1 above)
Comment (planctus) [reply src/implementations/twig/storybook/ec/.storybook/main.js]: not sure when you left this comment, if it's before the conversation we had on teams i guess we can close this.. :)

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 3]
```diff
@@ -1,40 +1,58 @@
...
+      category: 'Hidden content',
+    },
+    control: {
+      type: 'text',
+    },
+  };
+});
 
-  getExtraKnobs(data);
+const prepareDataWithArgs = (data, args) => {
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js]: I am not sure whether the change in name of this function matches what it actually does.
From the pull request for controls in the button component I put a comment about it. If we rename the function, which i think we should do, it should be related to specs, not args.
We might need to align whether this function will be the place where defaults are applied. 
At any case, it will be useful to have consistent naming and operation in this function throughout all stories.

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 4]
```diff
@@ -1,40 +1,58 @@
...
-      tabLabels.optional
-    );
-  });
+const argTypes = {};
+demoData.items.forEach((item, i) => {
+  argTypes[`toggle${i + 1}`] = {
+    name: `toggle ${i + 1}`,
+    type: { name: 'string', required: true },
+    defaultValue: item.toggle.label,
+    description: `Text of the accordion toggler`,
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js]: The plural of toggle is either toggles or toggle switches, but not togglers :)

File: src/implementations/twig/components/accordion/accordion.story.js (hunk 2 above)
Comment (kalinchernev) [reply src/implementations/twig/components/accordion/accordion.story.js:67]: I like the approach of disable-first and remove-when-done approach here

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 5]
```diff
@@ -1,40 +1,58 @@
...
-      tabLabels.required
-    );
-    item.level = select(
-      `items[${index}].level`,
-      levels,
-      level,
-      tabLabels.optional
-    );
-  });
+const argTypes = {};
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js]: Let's put argTypes generation logic in a function in the same way we keep logic for applying defaults based on specs data and have something like `Default.argTypes = getArgTypes();` Others later if having variants will be `Variant.argTypes = getArgTypesVariant();`

**Review by kalinchernev — CHANGES_REQUESTED** [review]

File: src/implementations/twig/components/accordion/accordion.story.js (hunk 4 above)
Comment (planctus) [reply src/implementations/twig/components/accordion/accordion.story.js]: true!! :)

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 6]
```diff
@@ -1,40 +1,62 @@
...
+      },
+    };
   });
 
-  getExtraKnobs(data);
+  return argTypes;
+};
+
+const applySpecs = (data, args) => {
+  demoData.items.forEach((item, i) => {
```
Comment (papegaill) [inline src/implementations/twig/components/accordion/accordion.story.js]: `data` instead of `demoData` ?

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 7]
```diff
@@ -1,40 +1,62 @@
...
+    item.toggle.icon.path = defaultSprite;
+    item.toggle.icon.type = 'general';
+    item.toggle.icon.size = 's';
+    if (i === 0) {
+      item.toggle.icon.name = 'digital';
```
Comment (papegaill) [inline src/implementations/twig/components/accordion/accordion.story.js]: I don't understand this condition for the toggle `icon.name`.. as it is a `plus` icon for every toggle?

File: src/implementations/twig/components/accordion/accordion.story.js (hunk 6 above)
Comment (planctus) [reply src/implementations/twig/components/accordion/accordion.story.js]: it's a parameter name, we pass demoData to it, which is how we call what we import from the specs when we call the function..

File: src/implementations/twig/components/accordion/accordion.story.js (hunk 7 above)
Comment (planctus) [reply src/implementations/twig/components/accordion/accordion.story.js]: ops, funny, i guess that i was looking at the old-accordion specs when i did this..thanks for noticing it :)

Comment (planctus) [reply src/implementations/twig/components/accordion/accordion.story.js]: i fixed it :)

File: src/implementations/twig/components/accordion/accordion.story.js (hunk 6 above)
Comment (kalinchernev) [reply src/implementations/twig/components/accordion/accordion.story.js]: I agree with Alexis, `demoData` should be `data`.
The parameter is `data`, the return value is `data`, whereas `demoData` is a variable from the module's scope, not the `applySpecs()` scope. It's a code smell to mutate variable from outer scope, as well as passing around parameter as a return without any modifications to it.

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 8]
```diff
@@ -1,40 +1,53 @@
...
+        type: 'text',
+      },
+    };
   });
 
-  getExtraKnobs(data);
+  return argTypes;
+};
+
+const applySpecs = (data, args) => {
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js]: With apologies for the change of mind, but this function is `prepareData()` rather than applying specs. I didn't pay attention in an earlier comment what's happening in the function.

**Review by kalinchernev — CHANGES_REQUESTED** [review]
Summary of change requests:
- rename `applySpecs` to `prepareData`
- change `demoData` to `data` for the scope of `prepareData` to avoid scope creep

```js
const prepareData = (data, args) => {
  data.items.forEach((item, i) => {
    item.toggle.icon.path = defaultSprite;
    item.toggle.label = args[`toggle${i + 1}`];
    item.content = args[`content${i + 1}`];
  });

  return data;
};
```

File: .eslintrc.js [hunk 9]
```diff
@@ -22,5 +22,11 @@ module.exports = {
         'no-undef': 'off',
       },
     },
+    {
```
Comment (kalinchernev) [inline .eslintrc.js]: Please remove this ignore rule. This is not a solution to the problem at hand, but an illustration of the scope creep commented on earlier.

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 10]
```diff
@@ -1,40 +1,53 @@
-import { withKnobs, text, select, optionsKnob } from '@storybook/addon-knobs';
 import { withNotes } from '@ecl/storybook-addon-notes';
-import { getExtraKnobs, tabLabels } from '@ecl/story-utils';
 import withCode from '@ecl/storybook-addon-code';
 import defaultSprite from '@ecl/resources-ec-icons/dist/sprites/icons.svg';
-import demoData from '@ecl/specs-component-accordion/demo/data';
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js:6]: First, leave this be `demoData`, there is no issue with it.

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 11]
```diff
@@ -1,40 +1,53 @@
...
-      { display: 'inline-radio' },
-      tabLabels.required
-    );
-    item.level = select(
-      `items[${index}].level`,
-      levels,
-      level,
-      tabLabels.optional
-    );
+const getArgTypes = () => {
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js]: Two, change this function to accept a parameter: `data`. This will bind the `data` in `getArgTypes` function scope, the contents of the curly brackets after the function name.

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 12]
```diff
@@ -43,8 +56,16 @@ export default {
   title: 'Components/Accordion',
 };
 
-export const Default = () => accordion(prepareAccordion(demoData));
-
+export const Default = (args) => accordion(prepareData(data, args));
+Default.argTypes = getArgTypes();
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js]: Three, invoke `getArgTypes()` with a value for the input parameter. In the scope of the module `accordion.story.js` it's going to be `demoData`, therefore

```javascript
Default.argTypes = getArgTypes(demoData);
```

Inside the `getArgTypes`, `demoData` will be replaced with `data` and there will be no shadowing or scope creep.

The main benefit: you can extract the function out of the module at any time without any issues. But you also make the logic flow clear and do not risk accidental changes of the data.

File: src/implementations/twig/components/accordion/accordion.story.js [hunk 13]
```diff
@@ -43,8 +56,16 @@ export default {
...
 Default.storyName = 'default';
-Default.parameters = { notes: { markdown: notes, json: demoData } };
```
Comment (kalinchernev) [inline src/implementations/twig/components/accordion/accordion.story.js:49]: Lastly, this is also `demoData` from module scope.

File: src/playground/addons/storybook-addon-system-switcher/systemSwitcher.jsx [hunk 14]
```diff
@@ -3,10 +3,9 @@ import { IconButton, Separator } from '@storybook/components';
 
 const Switcher = () => {
   const queryString = window.location.search;
-  let system = window.location.pathname.split('/').filter((p) => p)[1];
-  if (!system) {
-    system = process.env.STORYBOOK_SYSTEM ? 'eu' : 'ec';
-  }
+  const system =
```
Comment (kalinchernev) [inline src/playground/addons/storybook-addon-system-switcher/systemSwitcher.jsx]: Noticing a change here in the switcher. We need to have https://citnet.tech.ec.europa.eu/CITnet/jira/browse/FRONT-1838 soon in order to avoid difference in system extraction logic.

Here's the way I did it in the code addon, taking env variable first and first system-specific match from path as a second.

```javascript
const system = (
  window.location.pathname
    .split('/')
    .find((part) => part === 'ec' || part === 'eu') ||
  process.env.STORYBOOK_SYSTEM
).toLowerCase();
```

**Review by kalinchernev — CHANGES_REQUESTED** [review]

**Review by papegaill — APPROVED** [review]
